src/Symfony/Component/Security/Core/Tests/Authorization/Strategy/ScoringStrategyTest.php

```diff
@@ -0,0 +1,159 @@
+use Symfony\Component\Security\Core\Exception\LogicException;
+
+class ScoringStrategyTest extends TestCase
+{
+    /**
+     * @dataProvider provideStrategyTests
+     *
+     * @param VoterInterface[] $voters
+     */
+    final public function testGetDecision(AccessDecisionStrategyInterface $strategy, array $voters, AccessDecision $expected)
+    {
+        $token = $this->createMock(TokenInterface::class);
+        $manager = new AccessDecisionManager($voters, $strategy);
+
+        $this->assertEquals($expected, $manager->getDecision($token, ['ROLE_FOO']));
+    }
+
+    public static function provideStrategyTests(): iterable
+    {
+        $strategy = new ScoringStrategy();
+
+        yield [$strategy, [
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_ABSTAIN),
+        ], self::getAccessDecision(false, [
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_ABSTAIN, false],
+        ], 'score = -1'
+        )];
+
+        yield [$strategy, [
+            self::getVoterWithVoteObject(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_ABSTAIN),
+        ], self::getAccessDecision(false, [
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_ABSTAIN, false],
+        ], 'score = -1'
+        )];
+
+        yield [$strategy, [
+            self::getVoterWithVoteObjectAndScoring(5),
+        ], self::getAccessDecision(true, [
+            [5, true],
+        ], 'score = 5'
+        )];
+
+        yield [$strategy, [
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoterWithVoteObjectAndScoring(VoterInterface::ACCESS_ABSTAIN),
+            self::getVoterWithVoteObjectAndScoring(2),
+        ], self::getAccessDecision(false, [
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_ABSTAIN, true],
+            [2, true],
+        ], 'score = -1'
+        )];
+
+        yield [$strategy, [
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoter(VoterInterface::ACCESS_DENIED),
+            self::getVoterWithVoteObjectAndScoring(5),
+        ], self::getAccessDecision(true, [
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_DENIED, false],
+            [VoterInterface::ACCESS_DENIED, false],
+            [5, true],
+        ], 'score = 1'
+        )];
+    }
+
+
+    final protected static function getVoter(int $vote): VoterInterface
```

Comment:
classes in the "Tests" (note the "s") namespace shouldn't use the final modifier, and methods should be made private if possible.

```suggestion
    private static function getVoter(int $vote): VoterInterface
```